torchani/data/new.py

```diff
@@ -0,0 +1,449 @@
+            species = np.array([species_dict[x] for x in molecule['species']])
+            self.data_species += list(np.tile(species, (num_conformations, 1)))
+            # energies
+            self.data_energies += list(molecule['energies'].reshape((-1, 1)).astype(np.float32))
+            if transform:
+                self_energies = np.array(sum([self_energies_dict[x] for x in molecule['species']]))
+                self.data_self_energies += list(np.tile(self_energies, (num_conformations, 1)).astype(np.float32))
+            # other properties
+            if anidata_size > 5 and pkbar_installed is not None:
+                pbar.update(i)
+
+        if transform:
+            self.data_energies = np.array(self.data_energies) - np.array(self.data_self_energies)
+            import gc
+            del self.data_self_energies
+            del self_energies
+            gc.collect()
+
+        self.batch_size = batch_size
+        self.length = (len(self.data_species) + self.batch_size - 1) // self.batch_size
+        self.device = device
+
+        self.shuffled_index = np.arange(len(self.data_species))
+        np.random.shuffle(self.shuffled_index)
+
+        self.bar = bar
+        self.test_bar_max = test_bar_max
+        if not self.bar:
+            self.bar = 1000
+        if self.test_bar_max:
+            self.__getitem__(0, test_bar_max=self.test_bar_max)
+
+        anidata.cleanup()
+        import gc
+        del num_conformations
+        del species
+        del anidata
+        gc.collect()
+
+    @functools.lru_cache(maxsize=None)
+    def __getitem__(self, index, test_bar_max=None):
+
+        if index >= self.length:
+            raise IndexError()
+
+        batch_indices = slice(index * self.batch_size, (index + 1) * self.batch_size)
+        batch_indices_shuffled = self.shuffled_index[batch_indices]
+
+        batch_species = [self.data_species[i] for i in batch_indices_shuffled]
+        batch_coordinates = [self.data_coordinates[i] for i in batch_indices_shuffled]
+        batch_energies = [self.data_energies[i] for i in batch_indices_shuffled]
+
+        datas = [batch_species, batch_coordinates, batch_energies]
+
+        # get sort index
+        num_atoms_each_mole = [b.shape[0] for b in batch_species]
+        atoms = torch.tensor(num_atoms_each_mole, dtype=torch.int32)
+        sorted_atoms, sorted_atoms_idx = torch.sort(atoms, descending=False)
+
+        # sort each batch of data
+        for i, d in enumerate(datas):
+            datas[i] = self.sort_list_with_index(d, sorted_atoms_idx.numpy())
+
+        # get chunk size
+        output, count = torch.unique(atoms, sorted=True, return_counts=True)
+        counts = torch.cat((output.unsqueeze(-1).int(), count.unsqueeze(-1).int()), dim=-1)
+        chunk_size_list, chunk_max_list = split_to_chunks(counts, bar=self.bar * self.batch_size * 20)
+
+        # optimize bar, if test_bar_max is not None
+        if test_bar_max:
```

Comment:
Is this for debugging? It is OK (actually good) to have code that allows developers to easily print that information. But don't make this as an argument for an internal method. Use `self.debug` instead. Also, this function is long, please move the debugging code to a separate method.

Reply:
it's not debuging, it needs to used for every h5 file when user want to get a reasonable bar.
more detail:  https://github.com/aiqm/torchani/pull/284#issuecomment-519657353

Reply:
Then create a separate method `find_threshold`, not hacking in this method

Reply:
please specify how to use this find_threshold function.
CacheDataset is ok to have that.
but ShuffleDataset is a function, which use data from TorchData, TorchData should not be visible to user.